custom_components/zwave_mqtt/__init__.py

```diff
@@ -89,31 +94,52 @@ def async_node_changed(node):
     @callback
     def async_node_removed(node):
         _LOGGER.debug("[NODE REMOVED] node_id: %s", node.id)
+        data_nodes.pop(node.id)
+        # node added/removed events also happen on (re)starts of hass/mqtt/ozw
+        # cleanup device/entity registry if we know this node is permanently deleted
+        # entities itself are removed by the values logic
+        if node.id in removed_nodes:
+            hass.async_add_job(handle_remove_node(hass, node.id))
+
+    def async_instance_event(message):
+        event = message["event"]
+        event_data = message["data"]
+        _LOGGER.debug("[INSTANCE EVENT]: %s - data: %s", event, event_data)
+        # The actual removal action of a Z-Wave node is reported as instance event
+        # Only when this event is detected we cleanup the device and entities from hass
+        if event == "removenode" and "Node" in event_data:
```

Comment:
Should we break out constants for the event event data names?

Reply:
@MartinHjelmare yeah, I spotted a lot more places with hardcoded strings which could be moved to constants. But the best thing to do would be to add another enum in the library for that. I'll submit another PR soon for that improvement. I've already moved all references to CommandClasses to the lib in an earlier PR.

Are you OK to leave this as-is for now to be improved later all-at-once ?

Reply:
Yes, sounds very good!